async-nats/src/jetstream/consumer/pull.rs

```diff
@@ -75,59 +73,26 @@ impl Consumer<Config> {
         .await
     }
 
-    pub fn process(
-        &'_ mut self,
-        batch: usize,
-    ) -> impl Stream<Item = Result<Option<JetStreamMessage>, Error>> + '_ {
-        try_stream! {
-            let inbox = self.context.client.new_inbox();
-            let mut sub = self.context.client.subscribe(inbox.clone()).await?;
+    pub fn process(&'_ mut self, batch: usize) -> Result<Process, Error> {
+        let context = self.context.clone();
+        let subject = format!(
+            "{}.CONSUMER.MSG.NEXT.{}.{}",
+            self.context.prefix, self.info.stream_name, self.info.name
+        );
 
-            self.request_batch(BatchConfig {
-                batch: batch,
-                expires: None,
-                no_wait: false,
-                ..Default::default()
-            }, inbox.clone())
-            .await?;
-            let mut remaining = batch;
-            while let Some(mut message) = sub.next().await {
-                message.status = match message.status.unwrap_or(StatusCode::OK) {
-                    StatusCode::TIMEOUT => {
-                        remaining = batch;
-                        self.request_batch(BatchConfig {
-                            batch: batch,
-                            expires: None,
-                            no_wait: false,
-                            ..Default::default()
-                        }, inbox.clone())
-                        .await?;
-                        continue;
-                    }
-                    StatusCode::IDLE_HEARBEAT => {
-                        continue;
-                    }
-                    status => Some(status),
-                };
-
-                remaining -= 1;
-                yield Some(JetStreamMessage {
-                    context: self.context.clone(),
-                    message: message,
-                });
-
-                if remaining == 0 {
-                    self.request_batch(BatchConfig {
-                        batch: batch,
-                        expires: None,
-                        no_wait: false,
-                        ..Default::default()
-                    }, inbox.clone())
-                    .await?;
-                    remaining = batch;
-                }
-            }
-        }
+        let request = serde_json::to_vec(&BatchConfig {
+            batch,
+            ..Default::default()
```

Comment:
> This PR isa nice codebase improvment, but is not doing what process() should be doing:

Steam of streams, we skipped fundamentals on the first pass.
The baseline `IntoStream` can be built on top of any iterator, be it sequential or concurrent.

> Process should not expose underlying Batch operations.

Why not?  It's the requests that are fallible and require results and there is no reason to hide the low level access.

You can use a combinator like `flatten` (e.g `consumer.sequence(...).flatten()` and/or provide one on consumer directly that flattens if it can be done in an optimized way that can't be chained.

> Sending new Batch should not be sent only after previous one was drained, but in configurable way, optimized to be sent at given point of processing current batch.

Out of scope of this iterator, but yes we should have iterators for that. And measure the performance of each heuristic we come up with.

> So that does in fact remove async-stream dependency, but exposes internals that should not be exposed and removes important performance related feature.

Again don't consider the fact that fallible pull requests are being made internal. I, (the user) should be able to make my own iterator with better throttling heuristics that's better yours (us).

Maybe this is `Sequence`, and the initial implementation of `TryStream` (name tbd) for tomorrow's release will be `sequence().flatten()`?

Reply:
For the actual simple "the best async iterator of messages we have" case, I'm thinking:

```rust
trait TryStream {
  try_stream()
}

impl TryStream for PushConsuemr {} // infallible
impl TryStream for PullConsumer {} // fallible, but report fatal errors only.
```